beacon-chain/operations/blstoexec/pool.go

```diff
@@ -0,0 +1,135 @@
+package blstoexec
+
+import (
+	"math"
+	"sync"
+
+	"github.com/prysmaticlabs/prysm/v3/config/params"
+	types "github.com/prysmaticlabs/prysm/v3/consensus-types/primitives"
+	ethpb "github.com/prysmaticlabs/prysm/v3/proto/prysm/v1alpha1"
+)
+
+type listNode struct {
+	value *ethpb.SignedBLSToExecutionChange
+	prev  *listNode
+	next  *listNode
+}
+
+type changesList struct {
+	first *listNode
+	last  *listNode
+	len   int
+}
+
+// PoolManager maintains pending and seen BLS-to-execution-change objects.
+// This pool is used by proposers to insert BLS-to-execution-change objects into new blocks.
+type PoolManager interface {
+	PendingBLSToExecChanges() []*ethpb.SignedBLSToExecutionChange
+	BLSToExecChangesForInclusion() []*ethpb.SignedBLSToExecutionChange
+	InsertBLSToExecChange(change *ethpb.SignedBLSToExecutionChange)
+	MarkIncluded(change *ethpb.SignedBLSToExecutionChange)
+}
+
+// Pool is a concrete implementation of PoolManager.
+type Pool struct {
+	lock    sync.RWMutex
+	pending changesList
+	m       map[types.ValidatorIndex]*listNode
+}
+
+// NewPool returns an initialized pool.
+func NewPool() *Pool {
+	return &Pool{
+		pending: changesList{},
+		m:       make(map[types.ValidatorIndex]*listNode),
+	}
+}
+
+// PendingBLSToExecChanges returns all objects from the pool.
+func (p *Pool) PendingBLSToExecChanges() []*ethpb.SignedBLSToExecutionChange {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+
+	result := make([]*ethpb.SignedBLSToExecutionChange, p.pending.len)
+	node := p.pending.first
+	for i := 0; node != nil; i++ {
+		result[i] = node.value
+		node = node.next
+	}
+	return result
+}
+
+// BLSToExecChangesForInclusion returns objects that are ready for inclusion at the given slot.
+// This method will not return more than the block enforced MaxBlsToExecutionChanges.
+func (p *Pool) BLSToExecChangesForInclusion() []*ethpb.SignedBLSToExecutionChange {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+
+	length := int(math.Min(float64(params.BeaconConfig().MaxBlsToExecutionChanges), float64(p.pending.len)))
+	result := make([]*ethpb.SignedBLSToExecutionChange, length)
+	node := p.pending.first
+	for i := 0; node != nil && i < length; i++ {
```

Comment:
This loop feels almost the same except for the middle conditional. The functions are very similar and not sure if we should have it as a separate one. If the intention of pending vs. awaiting inclusion is to cap the amounts needed in the latter, perhaps we only need the latter. How would the former be used in the codebase?